### PR TITLE
Decorator to check for unknown keyword arguments

### PR DESCRIPTION
**Pull request by jpakkane**
Github was acting weird, it could not see new commits. Thus I had to close and reopen this.

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -1214,6 +1213,97 @@ def get_cross_property_method(self, args, kwargs):
...
+lang_arg_kwargs = set(['c_args',
+                       'cpp_args',
+                       'd_args',
+                       'fortran_args',
+                       'java_args',
+                       'objc_args',
+                       'objcpp_args',
+                       'rust_args',
+                       'vala_args',
+                   ])
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -1214,6 +1213,97 @@ def get_cross_property_method(self, args, kwargs):
...
+                          'install_dir',
+                          'name_prefix',
+                          'name_suffix',
+                          'native',
+                          'objects',
+                          'override_options',
+                          'pic',
+                          'sources',
+                          'vs_module_defs',
+                      ])
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -1214,6 +1213,97 @@ def get_cross_property_method(self, args, kwargs):
...
+                    'jar': jar_kwargs,
+                    'project': {'version', 'meson_version', 'default_options', 'license', 'subproject_dir'},
+                    'run_target': {'command', 'depends'},
+                    'shared_library': shlib_kwargs,
+                    'shared_module': shmod_kwargs,
+                    'static_library': stlib_kwargs,
+                    'subproject': {'version', 'default_options'},
+                    'test': {'args', 'env', 'is_parallel', 'should_fail', 'timeout', 'workdir', 'suite'},
+                    'vcs_tag': {'input', 'output', 'fallback', 'command', 'replace_string'},
+}
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: mesonbuild/modules/__init__.py [hunk 4]
```diff
@@ -2,10 +2,26 @@
 
 from .. import build
 from .. import dependencies
+from .. import mlog
 from ..mesonlib import MesonException
+from ..interpreterbase import permittedKwargs, noKwargs
```
Comment (ignatenkobrain) [inline mesonbuild/modules/__init__.py]: __[flake8]__
* [F401] '..interpreterbase.noKwargs' imported but unused
* [F401] '..interpreterbase.permittedKwargs' imported but unused

File: mesonbuild/interpreter.py [hunk 5]
```diff
@@ -1214,6 +1213,97 @@ def get_cross_property_method(self, args, kwargs):
                 return args[1]
             raise InterpreterException('Unknown cross property: %s.' % propname)
 
+pch_kwargs = set(['c_pch', 'cpp_pch'])
```
Comment (QuLogic) [inline mesonbuild/interpreter.py]: Some more set literals.

File: mesonbuild/interpreter.py [hunk 6]
```diff
@@ -1214,6 +1213,97 @@ def get_cross_property_method(self, args, kwargs):
...
+                          'name_suffix',
+                          'native',
+                          'objects',
+                          'override_options',
+                          'pic',
+                          'sources',
+                          'vs_module_defs',
+                      ])
+
+build_target_common_kwargs = set()
```
Comment (QuLogic) [inline mesonbuild/interpreter.py]:
```python
build_target_common_kwargs = (
    buildtarget_kwargs |
    lang_arg_kwargs |
    pch_kwargs |
    vala_kwargs |
    rust_kwargs |
    cs_kwargs
)
```

File: mesonbuild/interpreter.py [hunk 7]
```diff
@@ -1214,6 +1213,97 @@ def get_cross_property_method(self, args, kwargs):
...
+build_target_common_kwargs.update(lang_arg_kwargs)
+build_target_common_kwargs.update(pch_kwargs)
+build_target_common_kwargs.update(vala_kwargs)
+build_target_common_kwargs.update(rust_kwargs)
+build_target_common_kwargs.update(cs_kwargs)
+
+exe_kwargs = set()
+exe_kwargs.update(build_target_common_kwargs)
+
+shlib_kwargs = set()
```
Comment (QuLogic) [inline mesonbuild/interpreter.py]: shlibs_kwargs = build_target_common_kwargs | {'version', 'soversion'}

File: mesonbuild/interpreter.py (hunk 3 above)
Comment (QuLogic) [inline mesonbuild/interpreter.py]: A single four-space indent for the entire thing seems much cleaner to me.

File: mesonbuild/modules/__init__.py (hunk 4 above)
Comment (ignatenkobrain) [inline mesonbuild/modules/__init__.py]: __[flake8]__
* [F401] '..interpreterbase.permittedKwargs' imported but unused
* [F401] '..interpreterbase.noKwargs' imported but unused

**Pull request merged**